File: languages/csharp/assessments/04-expert/solutions/main.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

sealed class DataSet
{
    public DataSet(List<int> values)
    {
        Values = values;
    }

    public List<int> Values { get; }
}

sealed class Summary
{
    public long Total { get; set; }
    public int Minimum { get; set; } = int.MaxValue;
    public int Maximum { get; set; } = int.MinValue;
}

class Program
{
    static void Main()
    {
        DataSet data = new DataSet(new List<int> { 12, 7, 25, 4, 31, 19, 2, 45, 18, 9, 27, 6 });

        int workerCount = 3;
        int chunkSize = (data.Values.Count + workerCount - 1) / workerCount;
        Summary globalSummary = new Summary();
        object gate = new object();

        Task[] tasks = new Task[workerCount];

        for (int workerId = 0; workerId < workerCount; workerId++)
        {
            int capturedWorkerId = workerId;
            int begin = capturedWorkerId * chunkSize;
            int end = Math.Min(begin + chunkSize, data.Values.Count);

            tasks[capturedWorkerId] = Task.Run(() =>
            {
                if (begin >= end)
                {
                    return;
                }

                long localTotal = 0;
                int localMin = data.Values[begin];
                int localMax = data.Values[begin];

                for (int index = begin; index < end; index++)
                {
                    int value = data.Values[index];
                    localTotal += value;

                    if (value < localMin)
                    {
                        localMin = value;
                    }

                    if (value > localMax)
                    {
                        localMax = value;
                    }
                }

                lock (gate)
                {
                    Console.WriteLine($"Worker {capturedWorkerId} partial sum: {localTotal}");
                    globalSummary.Total += localTotal;
                    if (localMin < globalSummary.Minimum)
                    {
                        globalSummary.Minimum = localMin;
                    }

                    if (localMax > globalSummary.Maximum)
                    {
                        globalSummary.Maximum = localMax;
                    }
                }
            });
        }

        Task.WaitAll(tasks);

        Console.WriteLine();
        Console.WriteLine("Final summary:");
        Console.WriteLine($"Total: {globalSummary.Total}");
        Console.WriteLine($"Minimum: {globalSummary.Minimum}");
        Console.WriteLine($"Maximum: {globalSummary.Maximum}");
    }
}

File: languages/python/assessments/04-expert/solutions/main.py
from __future__ import annotations

import threading


class Summary:
    def __init__(self) -> None:
        self.total = 0
        self.minimum = float("inf")
        self.maximum = float("-inf")


def main() -> None:
    data = [12, 7, 25, 4, 31, 19, 2, 45, 18, 9, 27, 6]

    worker_count = 3
    chunk_size = (len(data) + worker_count - 1) // worker_count

    summary = Summary()
    gate = threading.Lock()
    threads: list[threading.Thread] = []

    def worker(worker_id: int, start: int, stop: int) -> None:
        if start >= stop:
            return

        local_values = data[start:stop]
        local_total = sum(local_values)
        local_min = min(local_values)
        local_max = max(local_values)

        with gate:
            print(f"Worker {worker_id} partial sum: {local_total}")
            summary.total += local_total
            summary.minimum = min(summary.minimum, local_min)
            summary.maximum = max(summary.maximum, local_max)

    for worker_id in range(worker_count):
        begin = worker_id * chunk_size
        end = min(begin + chunk_size, len(data))
        thread = threading.Thread(target=worker, args=(worker_id, begin, end))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    print()
    print("Final summary:")
    print(f"Total: {summary.total}")
    print(f"Minimum: {int(summary.minimum)}")
    print(f"Maximum: {int(summary.maximum)}")


if __name__ == "__main__":
    main()
